crates/uv-distribution/src/metadata/lowering.rs

```diff
@@ -48,18 +48,16 @@ impl LoweredRequirement {
             // We require that when you use a package that's part of the workspace, ...
             !workspace.packages().contains_key(&requirement.name)
                 // ... it must be declared as a workspace dependency (`workspace = true`), ...
-                || matches!(
-                    source,
-                    Some(Source::Workspace {
-                        // By using toml, we technically support `workspace = false`.
-                        workspace: true
-                    })
-                )
+                || source.as_ref().is_some_and(|source| source.iter().all(|source| {
+                    matches!(source, Source::Workspace { workspace: true, .. })
+                }))
                 // ... except for recursive self-inclusion (extras that activate other extras), e.g.
                 // `framework[machine_learning]` depends on `framework[cuda]`.
                 || &requirement.name == project_name;
         if !workspace_package_declared {
-            return Err(LoweringError::UndeclaredWorkspacePackage);
+            return Either::Left(std::iter::once(Err(
+                LoweringError::UndeclaredWorkspacePackage,
+            )));
```

Comment:
I find this slightly awkward: we return an iterator, but if the "overall" requirement fails, we use once with an error. Perhaps the return type should more truthfully be: `Result<impl Iterator<Item = Result<LoweredRequirement, LoweringError>>, LoweringError>`

Reply:
I suppose that's true -- this does only apply to workspace packages. I'd like @BurntSushi opinion on it too.

Reply:
Ok, I left it as an iterator for now... I may look into rewriting it. I feel like an iterator is actually a simpler API, despite being more complicated "internally".

Reply:
(This method is run for all dependencies declared in the workspace. It returns a single entry the vast majority of the time.)